cpp/include/cuspatial/detail/cartesian_product_iterator.cuh

```diff
@@ -0,0 +1,227 @@
+#include <thrust/iterator/transform_iterator.h>
+#include <thrust/pair.h>
+#include <utility>
+#include "utility/size_from_offsets.cuh"
+
+namespace cuspatial {
+namespace detail {
+
+struct cartesian_product_group {
+  int32_t idx;
+  int32_t size;
+  int32_t offset;
+};
+
+struct cartesian_product_group_index {
+  cartesian_product_group group_a;
+  cartesian_product_group group_b;
+  int32_t element_a_idx;
+  int32_t element_b_idx;
+};
+
+template <typename OffsetIteratorA,
+          typename OffsetIteratorB,
+          typename SizeIteratorLHS,
+          typename SizeIteratorRHS,
+          typename LookupIteratorLHS,
+          typename LookupIteratorRHS>
+struct cartesian_product_functor {
+  int32_t const num_elements_b;
+  SizeIteratorLHS const group_sizes_a;
+  SizeIteratorRHS const group_sizes_b;
+  OffsetIteratorA const group_offsets_a;
+  OffsetIteratorB const group_offsets_b;
+  LookupIteratorLHS const group_lookup_a;
+  LookupIteratorRHS const group_lookup_b;
+
+  __device__ cartesian_product_group_index operator()(int64_t const idx) const
+  {
+    // first dimension
+    int32_t group_lookup_idx_a = idx / num_elements_b;
+    auto group_idx_a           = *(group_lookup_a + group_lookup_idx_a);
+    auto group_size_a          = *(group_sizes_a + group_idx_a);
+    auto group_offset_a        = *(group_offsets_a + group_idx_a);
+    auto group_block_offset_a  = group_offset_a * num_elements_b;
+
+    // second dimension
+    int32_t group_lookup_idx_b = (idx - group_block_offset_a) / group_size_a;
+    auto group_idx_b           = *(group_lookup_b + group_lookup_idx_b);
+    auto group_size_b          = *(group_sizes_b + group_idx_b);
+    auto group_offset_b        = *(group_offsets_b + group_idx_b);
+    auto group_block_offset_b  = group_offset_b * group_size_a;
+
+    // relative index
+    int32_t relative_idx        = idx - group_block_offset_a - group_block_offset_b;
+    auto relative_element_idx_a = relative_idx % group_size_a;
+    auto relative_element_idx_b = relative_idx / group_size_a;
+
+    return cartesian_product_group_index{{group_idx_a, group_size_a, group_offset_a},
+                                         {group_idx_b, group_size_b, group_offset_b},
+                                         relative_element_idx_a,
+                                         relative_element_idx_b};
+  }
+};
+
+template <typename OffsetIterator>
+struct group_lookup_functor {
+  OffsetIterator group_offsets;
+  int32_t group_count;
+  __device__ int32_t operator()(int32_t element_idx) const
+  {
+    return thrust::distance(
+             group_offsets,
+             thrust::upper_bound(
+               thrust::seq, group_offsets, group_offsets + group_count, element_idx)) -
+           1;
+  }
+};
+
+/**
+ * @brief Makes an iterator which consecutively produces a cartesian product of two iterators.
+ *
+ * Pairs of elements grouped by offsets `A` and `B` are produced consecutively.
```

Comment:
Explaining what this iterator _does_ and why it's useful is actually very difficult to me, probably because I'm so close to the problem/solution/code. I will try to clarify, but the more suggestions the merrier.